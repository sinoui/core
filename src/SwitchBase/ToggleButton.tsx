import styled from 'styled-components';
import IconButton from '@sinoui/core/IconButton';
import getColorFromTheme from '@sinoui/core/utils/getColorFromTheme';
import { Theme } from '@sinoui/theme';

/**
 * 颜色
 */
const colorWrapper = (props: ToggleButtonProps) => {
  let color;
  if (!props.checked && !props.disabled) {
    color = getColorFromTheme(props.theme, 'textSecondary');
  } else if (props.disabled) {
    color = getColorFromTheme(props.theme, 'actionDisabled');
  } else {
    color = getColorFromTheme(props.theme, props.color);
  }
  return color;
};

export interface ToggleButtonProps {
  checked?: boolean;
  disabled?: boolean;
  color?: string;
  theme: Theme;
}

const ToggleButton = styled(IconButton).attrs(() => ({
  type: 'switch',
}))<ToggleButtonProps>`
  color: ${(props) => colorWrapper(props)};
  > .sinoui-icon,
  > .sinoui-svg-icon {
    font-size: 20px;
  }
  width: 20px;
  height: 20px;
  > .sinoui-icon-button__ripple-layout-dense {
    left: -6px;
    top: -6px;
  }
  > .sinoui-icon-button__ripple-layout {
    left: -14px;
    top: -14px;
  }
`;

export default ToggleButton;
