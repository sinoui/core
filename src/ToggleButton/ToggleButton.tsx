import classNames from 'classnames';
import { opacify } from 'polished';
import React from 'react';
import styled, { css } from 'styled-components';
import BaseButton from '../BaseButton';
import getColorFromTheme from '../utils/getColorFromTheme';

export interface ToggleButtonProps {
  /**
   * 是否被选中，默认为false
   */
  selected?: boolean;
  /**
   * 值
   */
  value: string;
  /**
   * 值变更时的回调函数
   */
  onChange?: (value: string) => void;
  /**
   * 子元素
   */
  children: React.ReactNode;
  /**
   * 指定颜色
   */
  color?: string;
  /**
   * 是否可用
   */
  disabled?: boolean;
  /**
   * 是否使用紧凑版
   */
  dense?: boolean;
}

const defaultButtonStyle = css<ToggleButtonProps>`
  ${({ theme }) => ({ ...theme.typography.button })};

  color: ${({ theme, disabled }) =>
    disabled ? theme.palette.text.disabled : 'rgba(0,0,0,0.54)'};
  user-select: none;
  transition: ${({ theme: { transitions } }) =>
    transitions.create(['color', 'background-color'], {
      easing: transitions.easing.easeInOut,
      duration: transitions.duration.shortest,
    })};
  padding: ${({ dense }) => (dense ? 7 : 11)}px;
  box-sizing: border-box;
  border: 1px solid ${({ theme }) => theme.palette.action.disabledBackground};
  border-radius: 4px;

  &:hover {
    background-color: ${({ theme, color }) =>
      opacify(
        theme.palette.action.hoverOpacity - 1,
        getColorFromTheme(theme, color) || theme.palette.primary.main,
      )};

    @media (hover: none) {
      background-color: transparent;
    }
  }
`;

const selectedStyle = css<ToggleButtonProps>`
  background-color: ${({ theme, color }) =>
    opacify(
      theme.palette.action.selectedOpacity - 1,
      getColorFromTheme(theme, color) || 'black',
    )};
  color: ${({ theme, color, disabled }) =>
    disabled
      ? theme.palette.text.disabled
      : getColorFromTheme(theme, color) || theme.palette.text.primary};

  &:hover {
    background-color: ${({ theme, color }) =>
      opacify(
        theme.palette.action.selectedOpacity - 1,
        getColorFromTheme(theme, color) || 'black',
      )};
    color: ${({ theme, color, disabled }) =>
      disabled
        ? theme.palette.text.disabled
        : getColorFromTheme(theme, color) || theme.palette.text.primary};

    @media (hover: none) {
      background-color: transparent;
    }
  }
`;

const ToggleButtonLayout = styled(BaseButton)<ToggleButtonProps>`
  ${defaultButtonStyle};
  ${({ selected }) => selected && selectedStyle};
`;

export default function ToggleButton(props: ToggleButtonProps) {
  const {
    children,
    onChange,
    selected,
    value,
    color,
    disabled,
    dense,
    ...rest
  } = props;

  const handleClcik = () => {
    if (onChange) {
      onChange(value);
    }
  };

  return (
    <ToggleButtonLayout
      {...rest}
      className={classNames('sinoui-toggle-button', {
        'sinoui-toggle-button--selected': selected,
        'sinoui-toggle-button--dense': dense,
      })}
      color={color}
      disabled={disabled}
      dense={dense}
      value={value}
      selected={selected}
      onClick={handleClcik}
    >
      {children}
    </ToggleButtonLayout>
  );
}
