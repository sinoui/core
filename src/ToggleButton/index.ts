import ToggleButton, { ToggleButtonProps } from './ToggleButton';

export default ToggleButton;

export type { ToggleButtonProps };
