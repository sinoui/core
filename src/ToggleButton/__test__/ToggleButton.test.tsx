/**
 * @jest-environment jsdom
 */
import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import '@testing-library/jest-dom';
import { render, cleanup, fireEvent, act } from '@testing-library/react';
import ToggleButton from '../ToggleButton';

afterEach(cleanup);

it('正确渲染切换按钮', () => {
  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <>
        <ToggleButton value="IOS">IOS</ToggleButton>
      </>
    </ThemeProvider>,
  );

  expect(tree).toMatchSnapshot();
});

it('渲染紧凑模式', () => {
  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <>
        <ToggleButton dense value="IOS">
          IOS
        </ToggleButton>
      </>
    </ThemeProvider>,
  );

  expect(tree).toMatchSnapshot();
});

it('选中状态', () => {
  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <>
        <ToggleButton selected value="IOS">
          IOS
        </ToggleButton>
        <ToggleButton color="primary" selected value="IOS">
          IOS
        </ToggleButton>
        <ToggleButton selected color="secondary" value="IOS">
          IOS
        </ToggleButton>
      </>
    </ThemeProvider>,
  );

  expect(tree).toMatchSnapshot();
});

it('渲染不同颜色的按钮', () => {
  const tree = renderer.create(
    <ThemeProvider theme={defaultTheme}>
      <ToggleButton value="IOS" color="primary">
        IOS
      </ToggleButton>
      <ToggleButton value="IOS" color="success">
        IOS
      </ToggleButton>
      <ToggleButton value="IOS" color="info">
        IOS
      </ToggleButton>
    </ThemeProvider>,
  );

  expect(tree).toMatchSnapshot();
});

describe('className', () => {
  it('基础渲染', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <ToggleButton value="check" data-testid="toggleButton">
          IOS
        </ToggleButton>
      </ThemeProvider>,
    );

    expect(getByTestId('toggleButton')).toHaveClass('sinoui-toggle-button');
  });

  it('选中状态', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <ToggleButton value="check" selected data-testid="toggleButton">
          IOS
        </ToggleButton>
      </ThemeProvider>,
    );

    expect(getByTestId('toggleButton')).toHaveClass(
      'sinoui-toggle-button--selected',
    );
  });

  it('紧凑模式', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <ToggleButton value="check" dense data-testid="toggleButton">
          IOS
        </ToggleButton>
      </ThemeProvider>,
    );

    expect(getByTestId('toggleButton')).toHaveClass(
      'sinoui-toggle-button--dense',
    );
  });

  it('点击按钮时，onchange被调用', () => {
    const onChange = jest.fn();
    const { getByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <ToggleButton
          value="check"
          onChange={onChange}
          dense
          data-testid="toggleButton"
        >
          IOS
        </ToggleButton>
      </ThemeProvider>,
    );

    const button = getByTestId('toggleButton');

    act(() => {
      fireEvent.click(button);
    });

    expect(onChange).toHaveBeenCalledWith('check');
  });
});
