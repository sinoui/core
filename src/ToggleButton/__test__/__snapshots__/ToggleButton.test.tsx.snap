// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`正确渲染切换按钮 1`] = `
<button
  className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ jsrPIs sinoui-toggle-button"
  onClick={[Function]}
  role="button"
  type="button"
  value="IOS"
>
  IOS
</button>
`;

exports[`渲染不同颜色的按钮 1`] = `
Array [
  <button
    className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ jsrPIs sinoui-toggle-button"
    color="primary"
    onClick={[Function]}
    role="button"
    type="button"
    value="IOS"
  >
    IOS
  </button>,
  <button
    className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ ixkGTa sinoui-toggle-button"
    color="success"
    onClick={[Function]}
    role="button"
    type="button"
    value="IOS"
  >
    IOS
  </button>,
  <button
    className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ fbeYvB sinoui-toggle-button"
    color="info"
    onClick={[Function]}
    role="button"
    type="button"
    value="IOS"
  >
    IOS
  </button>,
]
`;

exports[`渲染紧凑模式 1`] = `
<button
  className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ eaArLj sinoui-toggle-button sinoui-toggle-button--dense"
  onClick={[Function]}
  role="button"
  type="button"
  value="IOS"
>
  IOS
</button>
`;

exports[`选中状态 1`] = `
Array [
  <button
    className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ gLoBXw sinoui-toggle-button sinoui-toggle-button--selected"
    onClick={[Function]}
    role="button"
    selected={true}
    type="button"
    value="IOS"
  >
    IOS
  </button>,
  <button
    className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ esBoyA sinoui-toggle-button sinoui-toggle-button--selected"
    color="primary"
    onClick={[Function]}
    role="button"
    selected={true}
    type="button"
    value="IOS"
  >
    IOS
  </button>,
  <button
    className="sc-bdvvtL eqWEUT sinoui-base-button sc-gsDKAQ fdFOdg sinoui-toggle-button sinoui-toggle-button--selected"
    color="secondary"
    onClick={[Function]}
    role="button"
    selected={true}
    type="button"
    value="IOS"
  >
    IOS
  </button>,
]
`;
